### ISSUE TITLE
Join using does not work with a self join

### ISSUE THREAD

**Issue by dain** [prestodb/presto#4775]
```
presto:tiny> explain select * from orders join orders using (orderkey);
Query 20160314_204230_00004_guj48 failed: line 1:23: Relations appear more than once: [orders]
com.facebook.presto.sql.analyzer.SemanticException: line 1:23: Relations appear more than once: [orders]
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitJoin(StatementAnalyzer.java:1118)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitJoin(StatementAnalyzer.java:220)
    at com.facebook.presto.sql.tree.Join.accept(Join.java:87)
    at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.analyzeFrom(StatementAnalyzer.java:1721)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitQuerySpecification(StatementAnalyzer.java:1014)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitQuerySpecification(StatementAnalyzer.java:220)
    at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:125)
    at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitQuery(StatementAnalyzer.java:811)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitQuery(StatementAnalyzer.java:220)
    at com.facebook.presto.sql.tree.Query.accept(Query.java:103)
    at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
    at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:60)
    at com.facebook.presto.sql.analyzer.QueryExplainer.getLogicalPlan(QueryExplainer.java:129)
    at com.facebook.presto.sql.analyzer.QueryExplainer.getPlan(QueryExplainer.java:91)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.getQueryPlan(StatementAnalyzer.java:792)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitExplain(StatementAnalyzer.java:773)
    at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitExplain(StatementAnalyzer.java:220)
    at com.facebook.presto.sql.tree.Explain.accept(Explain.java:73)
    at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
    at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:60)
    at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:258)
    at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:244)
    at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:208)
    at com.facebook.presto.execution.QueuedExecution.lambda$start$259(QueuedExecution.java:68)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
explain select * from orders join orders using (orderkey)
```

**Comment by harunurhan** [comment]
`select * from orders o1 join orders o2 using (orderkey);` actually works.

It fails because of the check: [StatementAnalyzer.java#L1117](https://github.com/facebook/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L1117).

I think current behaviour is correct, since both MySQL and PostgreSQL don't allow it as well.

**Comment by KBP-TDC** [comment]
Is this SQL statement just another way to express NATURAL JOIN?

**Comment by ilfrin** [comment]
we could at least have a better error message in this case

**Comment by harunurhan** [comment]
@ilfrin 
Although we could have a better one, the current one is actually really similar to what MySQL (` Not unique table/alias: 'table_name'`) and PostgreSQL (`table name "table_name" specified more than once`) have.

**Comment by martint** [comment]
We should look into what the SQL spec says about this case.

**Comment by mattsfuller** [comment]
@KBP-TDC -- yes. NATURAL is shorthand for USING with the list of all common columns.

This error is more about specifying the same relation twice. It doesn't seem like it would be related to using. The same error would appear if you did select * from orders join orders on .... or select * from orders, orders where ...

The behavior is correct as far as my understanding of the SQL spec.

**Comment by martint** [comment]
I looked briefly and I couldn't find anything that says that's not allowed. I can't think of a reason why it shouldn't be allowed, either. Ambiguities would only matter if columns are referenced directly.

**Comment by martint** [comment]
> The declared type of the rows of the `<joined table>` is the row type RT defined by the sequence of (`<field name>`, `<data type>`) pairs indicated by the sequence of column descriptors of the `<joined table>` taken in order.

Also,

> [...] Every structural element, whether a column, a field, or an attribute, is primarily a name paired with a declared type. The elements of a structure are ordered. Elements in different positions in the same structure can have the same declared type. Sometimes the Syntax Rules forbid two elements at different ordinal positions from having equivalent names (particularly when created through an `<SQL schema statement>`). **When two elements with equivalent names are permitted, then the element name is ambiguous and attempts to reference either element by name will raise a syntax error.**

**Comment by harunurhan** [comment]
Thanks @martint.

In addition, according to this SQL spec, we should allow this:
`SELECT * FROM t1 x JOIN t2 x USING (a)` 

also this
`SELECT x.some-unique-name FROM t1 x JOIN t2 x USING (a)`
the column `some-unique-name` can belong to t1 or t2

**Comment by electrum** [comment]
Yes, JOIN USING needs to be fixed: #2311

**Comment by martint** [comment]
This seems to be working on master (0.160-SNAPSHOT)
